File: samples/scala-replicatedentity-examples/src/main/scala/com/example/replicated/set/domain/SomeSet.scala
package com.example.replicated.set.domain

import com.akkaserverless.scalasdk.replicatedentity.ReplicatedEntity
import com.akkaserverless.scalasdk.replicatedentity.ReplicatedEntityContext
import com.akkaserverless.scalasdk.replicatedentity.ReplicatedSet
import com.example.replicated.set
import com.google.protobuf.empty.Empty

// This class was initially generated based on the .proto definition by Akka Serverless tooling.
//
// As long as this file exists it will not be overwritten: you can maintain it yourself,
// or delete it so it is regenerated as needed.

/** A replicated entity. */
class SomeSet(context: ReplicatedEntityContext) extends AbstractSomeSet {

  /** Command handler for "Add". */
  def add(currentData: ReplicatedSet[String], addElement: set.AddElement): ReplicatedEntity.Effect[Empty] = 
    effects
      .update(currentData.add(addElement.element))
      .thenReply(Empty.defaultInstance)
  

  /** Command handler for "Remove". */
  def remove(currentData: ReplicatedSet[String], removeElement: set.RemoveElement): ReplicatedEntity.Effect[Empty] =
    effects
      .update(currentData.remove(removeElement.element))
      .thenReply(Empty.defaultInstance)

  /** Command handler for "Get". */
  def get(currentData: ReplicatedSet[String], getElements: set.GetElements): ReplicatedEntity.Effect[set.CurrentElements] =
    effects.reply(set.CurrentElements(currentData.elements.toSeq))

}

File: samples/java-protobuf-replicatedentity-examples/src/main/java/com/example/replicated/set/domain/SomeSet.java
package com.example.replicated.set.domain;

import kalix.javasdk.replicatedentity.ReplicatedEntityContext;
import kalix.javasdk.replicatedentity.ReplicatedSet;
import com.example.replicated.set.SomeSetApi;
import com.google.protobuf.Empty;

import java.util.List;
import java.util.stream.Collectors;

public class SomeSet extends AbstractSomeSet {
  @SuppressWarnings("unused")
  private final String entityId;

  public SomeSet(ReplicatedEntityContext context) {
    this.entityId = context.entityId();
  }

  // tag::update[]
  @Override
  public Effect<Empty> add(ReplicatedSet<String> set, SomeSetApi.AddElement command) {
    return effects()
        .update(set.add(command.getElement())) // <1>
        .thenReply(Empty.getDefaultInstance());
  }

  @Override
  public Effect<Empty> remove(ReplicatedSet<String> set, SomeSetApi.RemoveElement command) {
    return effects()
        .update(set.remove(command.getElement())) // <1>
        .thenReply(Empty.getDefaultInstance());
  }
  // end::update[]

  // tag::get[]
  @Override
  public Effect<SomeSetApi.CurrentElements> get(
      ReplicatedSet<String> set, SomeSetApi.GetElements command) {
    List<String> elements =
        set.elements().stream() // <1>
            .sorted()
            .collect(Collectors.toList());

    SomeSetApi.CurrentElements currentElements =
        SomeSetApi.CurrentElements.newBuilder().addAllElements(elements).build();

    return effects().reply(currentElements);
  }
  // end::get[]
}
